zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/adhocsubprocess/AdHocSubProcessActivityActivateProcessor.java

```diff
@@ -0,0 +1,239 @@
+      flowNodeProcessInstanceRecord
+          .setFlowScopeKey(adHocSubprocessElementInstance.getKey())
+          .setElementId(flowNodeToActivate.getId())
+          .setBpmnElementType(flowNodeToActivate.getElementType())
+          .setBpmnEventType(flowNodeToActivate.getEventType());
+
+      final long elementToActivateInstanceKey = keyGenerator.nextKey();
+      commandWriter.appendFollowUpCommand(
+          elementToActivateInstanceKey,
+          ProcessInstanceIntent.ACTIVATE_ELEMENT,
+          flowNodeProcessInstanceRecord);
+    }
+
+    stateWriter.appendFollowUpEvent(
+        command.getKey(), AdHocSubProcessActivityActivationIntent.ACTIVATED, command.getValue());
+
+    responseWriter.writeEventOnCommand(
+        command.getKey(),
+        AdHocSubProcessActivityActivationIntent.ACTIVATED,
+        command.getValue(),
+        command);
+  }
+
+  @Override
+  public ProcessingError tryHandleError(
+      final TypedRecord<AdHocSubProcessActivityActivationRecord> command, final Throwable error) {
+    rejectionWriter.appendRejection(command, RejectionType.INVALID_ARGUMENT, error.getMessage());
+    responseWriter.writeRejectionOnCommand(
+        command, RejectionType.INVALID_ARGUMENT, error.getMessage());
+
+    return ProcessingError.EXPECTED_ERROR;
+  }
+
+  private void validateDuplicateFlowNodes(
+      final TypedRecord<AdHocSubProcessActivityActivationRecord> command) {
+    final var hasDuplicates =
+        command.getValue().getFlowNodes().stream()
+                .map(AdHocSubProcessActivityActivationFlowNodeValue::getFlowNodeId)
+                .distinct()
+                .count()
+            != command.getValue().getFlowNodes().size();
+    if (hasDuplicates) {
+      throw new DuplicateFlowNodesException(
+          command.getValue().getFlowNodes().stream()
+              .map(AdHocSubProcessActivityActivationFlowNodeValue::getFlowNodeId)
+              .toList());
+    }
+  }
+
+  private boolean isAuthorized(
+      final TypedRecord<AdHocSubProcessActivityActivationRecord> command,
+      final ElementInstance adHocSubprocessElementInstance) {
+    final var authRequest =
+        new AuthorizationRequest(
+                command,
+                AuthorizationResourceType.PROCESS_DEFINITION,
+                PermissionType.UPDATE_PROCESS_INSTANCE,
+                adHocSubprocessElementInstance.getValue().getTenantId())
+            .addResourceId(adHocSubprocessElementInstance.getValue().getBpmnProcessId());
+    final var isAuthorized = authCheckBehavior.isAuthorized(authRequest);
+    if (isAuthorized.isLeft()) {
+      final var rejection = isAuthorized.getLeft();
+      final String errorMessage =
+          RejectionType.NOT_FOUND.equals(rejection.type())
+              ? AuthorizationCheckBehavior.NOT_FOUND_ERROR_MESSAGE.formatted(
+                  "modify a process instance",
+                  command.getValue().getAdHocSubProcessInstanceKey(),
```

Comment:
💭 Should this be the process instance key?

Reply:
I felt like this wasn't making sense to me at all. I then checked how authz is handled in `ResourceDeletionDeleteProcessor` and noticed that they just throw a `ForbiddenException` if the authz is rejected. So now that's what we do now 😄